File: src/main/java/com/nc5/generator/generator/BsGenerator.java
package com.nc5.generator.generator;

import com.nc5.generator.config.BillConfig;
import com.nc5.generator.template.TemplateContext;
import com.nc5.generator.template.TemplateEngine;
import com.nc5.generator.template.TemplateSelector;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;

/**
 * 业务逻辑代码生成器
 */
public class BsGenerator {

    private static final Logger logger = LoggerFactory.getLogger(BsGenerator.class);

    private TemplateEngine templateEngine;

    private String outputDir;

    public BsGenerator(String outputDir) {
        this.templateEngine = new TemplateEngine();
        this.outputDir = outputDir;
    }

    /**
     * 生成所有业务逻辑代码
     */
    public void generate(BillConfig billConfig) throws IOException {
        logger.info("开始生成业务逻辑代码");

        // 生成InsertAction
        generateInsertAction(billConfig);

        // 生成UpdateAction
        generateUpdateAction(billConfig);

        // 生成DeleteAction
        generateDeleteAction(billConfig);

        // 生成SaveAction
        generateSaveAction(billConfig);

        // 生成QueryAction
        generateQueryAction(billConfig);

        // 生成发布动作类
        generatePubActions(billConfig);

        logger.info("业务逻辑代码生成完成");
    }

    /**
     * 生成InsertAction
     */
    private void generateInsertAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getInsertActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/private/nc/bs/{module}/{billCode}/InsertAction.java
        String outputPath = String.format("%s/src/private/nc/bs/%s/%s/InsertAction.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower());

        writeToFile(outputPath, content);
        logger.info("生成InsertAction: {}", outputPath);
    }

    /**
     * 生成UpdateAction
     */
    private void generateUpdateAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getUpdateActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/private/nc/bs/{module}/{billCode}/UpdateAction.java
        String outputPath = String.format("%s/src/private/nc/bs/%s/%s/UpdateAction.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower());

        writeToFile(outputPath, content);
        logger.info("生成UpdateAction: {}", outputPath);
    }

    /**
     * 生成DeleteAction
     */
    private void generateDeleteAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getDeleteActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/private/nc/bs/{module}/{billCode}/DeleteAction.java
        String outputPath = String.format("%s/src/private/nc/bs/%s/%s/DeleteAction.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower());

        writeToFile(outputPath, content);
        logger.info("生成DeleteAction: {}", outputPath);
    }

    /**
     * 生成SaveAction
     */
    private void generateSaveAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getSaveActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/public/nc/action/{module}/{billCode}/N_{billCode}_SAVE.java
        String outputPath = String.format("%s/src/public/nc/action/%s/%s/N_%s_SAVE.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成SaveAction: {}", outputPath);
    }

    /**
     * 生成QueryAction
     */
    private void generateQueryAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getQueryActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/private/nc/bs/{module}/{billCode}/QueryAction.java
        String outputPath = String.format("%s/src/private/nc/bs/%s/%s/QueryAction.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower());

        writeToFile(outputPath, content);
        logger.info("生成QueryAction: {}", outputPath);
    }

    /**
     * 生成发布动作类
     */
    private void generatePubActions(BillConfig billConfig) throws IOException {
        logger.info("开始生成发布动作类");
        
        // 定义发布动作配置：动作名称、模板文件名
        String[][] pubActions = {
            {"DELETE", "PubAction_DELETE.vm"},
            {"EDIT", "PubAction_EDIT.vm"},
            {"FREEZE", "PubAction_Simple.vm"},
            {"UNFREEZE", "PubAction_Simple.vm"},
            {"WRITE", "PubAction_WRITE.vm"},
            {"WRITEBATCH", "PubAction_WRITEBATCH.vm"},
            {"SAVE", "PubAction_SAVE.vm"},
            {"APPROVE", "PubAction_Approve.vm"}
        };
        
        for (String[] action : pubActions) {
            String actionName = action[0];
            String templateFile = action[1];
            
            generatePubAction(billConfig, actionName, templateFile);
        }
        
        logger.info("发布动作类生成完成");
    }
    
    /**
     * 生成单个发布动作类
     */
    private void generatePubAction(BillConfig billConfig, String actionName, String templateFile) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);
        
        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getPubActionTemplate(billConfig, templateFile);
        
        String content = templateEngine.render(templatePath, context);
        
        // 输出路径: src/private/nc/bs/pub/action/N_{BILLCODE}_{ACTIONNAME}.java
        String outputPath = String.format("%s/src/private/nc/bs/pub/action/N_%s_%s.java",
                outputDir,
                billConfig.getClassName(),
                actionName);
        
        writeToFile(outputPath, content);
        logger.info("生成发布动作类: N_{}{}", billConfig.getClassName(), actionName);
    }

    /**
     * 写入文件
     */
    private void writeToFile(String filePath, String content) throws IOException {
        File file = new File(filePath);
        File parentDir = file.getParentFile();

        if (!parentDir.exists()) {
            parentDir.mkdirs();
        }

        FileUtils.writeStringToFile(file, content, "GBK");
    }
}

File: antd-ahug/src/core/generators/BsGenerator.ts
import type { BillConfig } from '../models';
import TemplateEngine from '../utils/templateEngine';
import path from 'path';

export class BsGenerator {
  private templateEngine: TemplateEngine;
  private outputDir: string;

  constructor(outputDir: string) {
    this.outputDir = outputDir;
    this.templateEngine = new TemplateEngine();
  }

  async generate(billConfig: BillConfig): Promise<void> {
    console.log('开始生成业务逻辑代码');

    await this.generateInsertAction(billConfig);
    await this.generateUpdateAction(billConfig);
    await this.generateDeleteAction(billConfig);
    await this.generateSaveAction(billConfig);
    await this.generateQueryAction(billConfig);
    await this.generatePubActions(billConfig);

    console.log('业务逻辑代码生成完成');
  }

  private async generateInsertAction(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'bs/common/InsertAction.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/private/nc/bs/${billConfig.module}/${billConfig.billCode.toLowerCase()}/InsertAction.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成InsertAction: ${outputPath}`);
  }

  private async generateUpdateAction(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'bs/common/UpdateAction.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/private/nc/bs/${billConfig.module}/${billConfig.billCode.toLowerCase()}/UpdateAction.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成UpdateAction: ${outputPath}`);
  }

  private async generateDeleteAction(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'bs/common/DeleteAction.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/private/nc/bs/${billConfig.module}/${billConfig.billCode.toLowerCase()}/DeleteAction.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成DeleteAction: ${outputPath}`);
  }

  private async generateSaveAction(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'bs/common/SaveAction.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/public/nc/action/${billConfig.module}/${billConfig.billCode.toLowerCase()}/N_${billConfig.billCode}_SAVE.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成SaveAction: ${outputPath}`);
  }

  private async generateQueryAction(billConfig: BillConfig): Promise<void> {
    let templatePath: string;
    if (billConfig.billType === 'archive') {
      templatePath = 'bs/by-type/archive/QueryAction_Archive.ejs';
    } else if (billConfig.billType === 'multi') {
      templatePath = 'bs/by-type/multi/QueryAction_Multi.ejs';
    } else {
      templatePath = 'bs/by-type/single/QueryAction_Single.ejs';
    }

    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const outputPath = path.join(
      this.outputDir,
      `src/private/nc/bs/${billConfig.module}/${billConfig.billCode.toLowerCase()}/QueryAction.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成QueryAction: ${outputPath}`);
  }

  private async generatePubActions(billConfig: BillConfig): Promise<void> {
    console.log('开始生成发布动作类');
    
    const pubActions = [
      { actionName: 'DELETE', templateFile: 'PubAction_DELETE.ejs' },
      { actionName: 'EDIT', templateFile: 'PubAction_EDIT.ejs' },
      { actionName: 'FREEZE', templateFile: 'PubAction_Simple.ejs' },
      { actionName: 'UNFREEZE', templateFile: 'PubAction_Simple.ejs' },
      { actionName: 'WRITE', templateFile: 'PubAction_WRITE.ejs' },
      { actionName: 'WRITEBATCH', templateFile: 'PubAction_WRITEBATCH.ejs' },
      { actionName: 'SAVE', templateFile: 'PubAction_SAVE.ejs' },
      { actionName: 'APPROVE', templateFile: 'PubAction_Approve.ejs' }
    ];
    
    for (const action of pubActions) {
      await this.generatePubAction(billConfig, action.actionName, action.templateFile);
    }
    
    console.log('发布动作类生成完成');
  }
  
  private async generatePubAction(billConfig: BillConfig, actionName: string, templateFile: string): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = `bs/common/${templateFile}`;
    const outputPath = path.join(
      this.outputDir,
      `src/private/nc/bs/pub/action/N_${billConfig.billCode}_${actionName}.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成发布动作类: N_${billConfig.billCode}${actionName}`);
  }
}

export default BsGenerator;
